src/pkg/artifact/dao/dao.go

```diff
@@ -251,131 +241,89 @@ func (d *dao) DeleteReferences(ctx context.Context, parentID int64) error {
-			WithMessage(`the value of "base" query can only be exact match value with "*"`)
+func filterBase(query *q.Query) selector.Option {
+	return func(builder *selector.Builder, cond *selector.Cond) error {
+		base, exist := query.GetValue("base")
+		if !exist {
+			builder.Where(
+				cond.Or(
+					cond.Exists("SELECT 1 FROM tag WHERE tag.artifact_id = T0.id"),
+					cond.NotExists("SELECT 1 FROM artifact_reference ref WHERE ref.child_id = T0.id"),
+				),
+			)
+
+			return nil
+		}
+
+		b, ok := base.(string)
+		if !ok || b != "*" {
+			return errors.New(nil).WithCode(errors.BadRequestCode).
+				WithMessage(`the value of "base" query can only be exact match value with "*"`)
+		}
+
+		return nil
 	}
-	// the base is specified as "*"
-	return qs, nil
 }
 
 // handle query string: q=tags=value q=tags=~value
-func setTagQuery(ctx context.Context, qs beegoorm.QuerySeter, query *q.Query) (beegoorm.QuerySeter, error) {
-	if query == nil || len(query.Keywords) == 0 {
-		return qs, nil
-	}
-	tags, exist := query.Keywords["tags"]
-	if !exist {
-		tags, exist = query.Keywords["Tags"]
+func filterTag(query *q.Query) selector.Option {
+	return func(builder *selector.Builder, cond *selector.Cond) error {
+		tags, exist := query.GetValue("tags", "Tags")
 		if !exist {
-			return qs, nil
+			return nil
 		}
-	}
 
-	// fuzzy match
-	f, ok := tags.(*q.FuzzyMatchValue)
-	if ok {
-		// get the id list first to avoid the sql injection
-		inClause, err := orm.CreateInClause(ctx, `SELECT DISTINCT art.id FROM artifact art
-			JOIN tag ON art.id=tag.artifact_id
-			WHERE tag.name LIKE ?`, "%"+orm.Escape(f.Value)+"%")
-		if err != nil {
-			return nil, err
-		}
-		qs = qs.FilterRaw("id", inClause)
-		return qs, nil
-	}
-	// exact match:
-	// "*" for listing tagged artifacts
-	// "nil" for listing untagged artifacts
-	// others for get the artifact with the specified tag
-	s, ok := tags.(string)
-	if ok {
-		if s == "*" {
-			qs = qs.FilterRaw("id", tagged)
-			return qs, nil
-		}
-		if s == "nil" {
-			qs = qs.FilterRaw("id", untagged)
-			return qs, nil
+		// fuzzy match
+		if f, ok := tags.(*q.FuzzyMatchValue); ok {
+			builder.Join("tag AS t", "t.artifact_id = T0.id", cond.InsensitiveContains("t.name", f.Value))
+			return nil
 		}
 
-		// get the id list first to avoid the sql injection
-		inClause, err := orm.CreateInClause(ctx, `SELECT DISTINCT art.id FROM artifact art
-			JOIN tag ON art.id=tag.artifact_id
-			WHERE tag.name = ?`, s)
-		if err != nil {
-			return nil, err
+		// exact match:
+		// "*" for listing tagged artifacts
+		// "nil" for listing untagged artifacts
+		// others for get the artifact with the specified tag
+		if s, ok := tags.(string); ok {
+			if s == "*" {
+				builder.Where(cond.Exists("SELECT 1 FROM tag WHERE tag.artifact_id = T0.id"))
+			} else if s == "nil" {
+				builder.Where(cond.NotExists("SELECT 1 FROM tag WHERE tag.artifact_id = T0.id"))
+			} else {
+				builder.Join("tag AS t", "t.artifact_id = T0.id", cond.Equal("t.name", s))
```

Comment:
Is the sql injection handled by the framework?

Reply:
`cond.Equal("t.name", s)` will generate string `t.name = ?`